File: 01_query_api/02_example.py
# 02_example_python.py
# Simple example of making an API request in Python
# Pairs with 02_example.R
# Tim Fraser

# This script shows how to:
# - Load an API key from a .env file
# - Make a GET request to an example API
# - Inspect the HTTP status code and JSON response

# 0. Setup #################################

## 0.1 Load Packages ############################

# !pip install requests python-dotenv  # run this once in your environment

import os  # for reading environment variables
import requests  # for making HTTP requests
from dotenv import load_dotenv  # for loading variables from .env

## 0.2 Load Environment Variables ################

# Load environment variables from the .env file in the project root
# This matches the behavior of readRenviron(".env") in 02_example.R
load_dotenv(".env")

# Get the API key from the environment
TEST_API_KEY = os.getenv("TEST_API_KEY")

## 1. Make API Request ###########################

# Execute query (JSONPlaceholder: no Cloudflare, works in all networks)
url = "https://jsonplaceholder.typicode.com/users/1"
headers = {"Accept": "application/json"}
if os.getenv("TEST_API_KEY"):
    headers["x-api-key"] = os.getenv("TEST_API_KEY")
response = requests.get(url, headers=headers)

## 2. Inspect Response ###########################

# View response status code (200 = success)
print(response.status_code)

# Only parse JSON when response is OK and looks like JSON (avoid crash when server returns HTML)
if response.status_code == 200 and response.text.strip() and not response.text.strip().startswith("<"):
    print(response.json())
else:
    print("API returned HTML or error.")
    print("Response preview:", (response.text[:120] + "...") if len(response.text) > 120 else response.text)


# Clear environment (optional in short scripts, but shown for parity
# with the R example that clears its workspace)
globals().clear()

File: 01_query_api/02_example.R
# 02_example_r.R

# This script can be run in R to demonstrate 
# how to make an API request

# Install if you haven’t yet
# install.packages(c(“httr2”, “jsonlite”)) 

# Execute query and save response as object
library(httr2)
library(jsonlite)

# Load environment variables from .env (optional)
if (file.exists(".env")) readRenviron(".env")
TEST_API_KEY = Sys.getenv("TEST_API_KEY")

# Create request (JSONPlaceholder: no Cloudflare, works in all networks)
req = request("https://jsonplaceholder.typicode.com/users/1") |>
  req_headers("User-Agent" = "Mozilla/5.0 (compatible; R-httr2/1.0)")
if (nzchar(TEST_API_KEY)) req = req_headers(req, `x-api-key` = TEST_API_KEY)
req = req_method(req, "GET")

# Execute request and print response
resp = req_perform(req)
cat("Status Code:", resp_status(resp), "\n")
data = resp_body_json(resp)
print(data)

# Clear environment
rm(list = ls())

# Exit
# q(save = "no")